src/org/rascalmpl/library/lang/html/IO.java

```diff
@@ -63,22 +89,118 @@ public IValue readHTMLFile(ISourceLocation file) {
             throw RuntimeExceptionFactory.io(factory.string(e.getMessage()));
         }
     }
+
+    /**
+     * Produces a HTML string output from an HTMLElement AST.
+     * 
+     * Why go through all the trouble of building a DOM? The only reason is compliance.
+     * Escapes, encodings, etc. all are maintained by these classes from org.w3c.dom.
+     */
+    public IString writeHTMLString(IConstructor cons) {
+        try {
+            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+            DocumentBuilder db = dbf.newDocumentBuilder();
+            // dbf.setSchema(SchemaFactory.newDefaultInstance().newSchema(new URL("http://www.w3.org/1999/xhtml")));
+            
+            Document doc = db.newDocument();
+            
+            Element elem = doc.createElement("html");
+            createDocument(doc, cons, elem);
+
+            TransformerFactory tFactory = TransformerFactory.newInstance();
+            Transformer transformer = tFactory.newTransformer();
+            DOMSource source = new DOMSource(elem.getFirstChild());
+            StringWriter writer = new StringWriter();
+            StreamResult result = new StreamResult(writer);
+            
+            transformer.transform(source, result);
+
+            return factory.string(writer.toString());
+        }
+        catch (ParserConfigurationException | TransformerException e) {
+            throw RuntimeExceptionFactory.io(e.getMessage());
+        }
+    }
+
+     /**
+     * Produces a HTML string output from an HTMLElement AST.
+     * 
+     * Why go through all the trouble of building a DOM? The only reason is compliance.
```

Comment:
does introduce scaling issues, so we shouldn't be generating big HTML trees :)

Reply:
memory consumption. This is a thing with large xml & json files, but I'm guessing nobody is gonna build such a big html.